src/cryptonote_core/cryptonote_core.cpp

```diff
@@ -589,38 +589,38 @@ namespace cryptonote
     std::vector<result> results(tx_blobs.size());
 
     tvc.resize(tx_blobs.size());
-    tools::task_region(m_threadpool, [&] (tools::task_region_handle& region) {
-      std::list<blobdata>::const_iterator it = tx_blobs.begin();
-      for (size_t i = 0; i < tx_blobs.size(); i++, ++it) {
-        region.run([&, i, it] {
-          results[i].res = handle_incoming_tx_pre(*it, tvc[i], results[i].tx, results[i].hash, results[i].prefix_hash, keeped_by_block, relayed, do_not_relay);
-        });
+    tools::threadpool::waitobj wobj;
+    wobj.num = 0;
+    std::list<blobdata>::const_iterator it = tx_blobs.begin();
+    for (size_t i = 0; i < tx_blobs.size(); i++, ++it) {
+      m_threadpool.submit(&wobj, [&, i, it] {
+        results[i].res = handle_incoming_tx_pre(*it, tvc[i], results[i].tx, results[i].hash, results[i].prefix_hash, keeped_by_block, relayed, do_not_relay);
+      });
+    }
```

Comment:
I liked the task_region semantics. How about a new helper object, which would automatically have a waitobj local, and wait on it at the end, similar to task_region ?

Reply:
ok, I'll take a look at that too

Reply:
really I wasn't too thrilled with the task_region stuff. particularly right here, where there would be two in a row. IMO it was cleaner to have a single waitobj here instead of two instances of task_regions.

Reply:
Could just add a destructor that does a wait(). So it will be automatic within a scope.

Reply:
The `task_region` didn't need to allocate memory, as this waiter doesn't either. Don't know why I implemented it like that; I have an alternate version for the mymonero backend I just modified which doesn't allocate and uses a single counter for tracking references. So it similar to this patch, but the waiter is automatically created when going into the region scope. The `task_region` was just something I took from academics - it clearly designated the "join" point of all the tasks.

Reply:
Oh the discussion wasn't about the allocations. Don't know why having multiple cheap to create objects on the stack is negative ... ? What would the academics say!

Reply:
sync variables are not always cheap to create and destroy. Mutexes on Windows can be quite slow to create.

Reply:
My comment was about having an automatic mass join - automatically does the right thing.

Reply:
I hadn't considered the runtime of creating the mutex / condition variable themselves but its an obvious bottleneck in hindsight. And I think avoiding those types requires some user-space CPU burning or arbitrary sleeping implementation. The original version had an explicit wait function that could be called to mitigate that runtime issue, although it changes the intended usage slightly.

Reply:
I just recalled more of my previous design rationale. The `thread_group` had an internal `boost::optional` so that the `mutex` / `condition_variable`, etc., did not have to be initialized if the thread count == 0 (they were often being created temporarily on the stack, unfortunately). The `task_region` had a `shared_ptr` in part to delay the creation of the `mutex` / `condition_variable` pair too. Why I didn't go with the `boost::optional` for `task_region` instead of the allocation I don't remember, other than it looks less "sketchy".

Anyway, a `boost::optional` inside of the `waiter` type is an easy method to delay construction of those synchronization types if the implementation frequently runs on the local thread due to hitting the `max` queue. I just wanted to point this out in case you noticed this while running perf tests / profiling.